nomad/fsm.go

```diff
@@ -1931,8 +1888,12 @@ func (s *nomadSnapshot) persistIndexes(sink raft.SnapshotSink,
 		idx := raw.(*state.IndexEntry)
 
 		// Write out a node registration
-		sink.Write([]byte{byte(IndexSnapshot)})
-		if err := encoder.Encode(idx); err != nil {
+		_, err := sink.Write([]byte{byte(IndexSnapshot)})
```

Comment:
(This is less for @krishicks and more for Future Us in case someone wonders why we made this change.)

I think it's worth leaving a review comment on this change since changing error checking behavior, even to make it more strict, in such critical path is always inherently risky.

My *assumption* is that the original code skipped checking the success of Write for 2 reasons:

1. There's likely *layers* of buffering involved in this write that make it so this call will never return an error.
2. Even *if* the Write could error, the nature of disk IO errors is such that the Encode call will always return an error as well, thus the end result is the same as if both errors were checked.

I think those assertions are safe, but there's very little benefit to taking that risk. We have no evidence simple error checking has ever had a remotely measurable impact on performance, and failing the function 1 line early in the off chance this Write can error is always functionally identical to the old assumed behavior outlined above.

So all of that to say +1 to adding the strict error checking. It's good to be idiomatic and linter friendly. Our SOP is to always comment when errors are ignored, so when lacking justification I think adding the checks back in is a good default choice.